In [1]:
# Explore data with pandas
import pandas as pd
import numpy as np

iowa_data = pd.read_csv('./data/train.csv')
iowa_data.describe(include=[np.number])

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

### Continous variables to explore
1. LotFrontage 
2. LotArea 
3. MasVnrArea 
4. BsmtFinSF1 
5. BsmtFinSF2 
6. BsmtUnfSF 
7. TotalBsmtSF 
8. 1stFlrSF 
9. 2ndFlrSF 
10. LowQualFinSF 
11. BsmtFullBath 
12. BsmtHalfBath 
13. FullBath 
14. HalfBath 
15. BedroomAbvGr 
16. KitchenAbvGr 
17. Fireplaces 
18. LivArea 
19. GarageArea
20. WoodDeckSF 
21. OpenPorchSF 
22. EnclosedPorch 
23. 3SsnPorch 
24. ScreenPorch
25. PoolArea 
26. MiscVal 
27. SalePrice 

## How to identify important features to include in model
---
### 1. Correlation Analysis
Compute a correlation matrix to see the correlation between each feature and the target.

Number should range between -1 and 1.
When the number is closer to either extreme, we can say there is a strong correlation.
As a rule of thumb for using this to pick features for the model we want to look for coefficients of 0.3 and above (inverse being -0.3 and below)


In [2]:
# Let use a correlation matrix to see the pearsone correlation coefficient between variables
corr_matrix = iowa_data.corr(numeric_only=True)
corr_matrix["SalePrice"].sort_values(ascending=False)

SalePrice        1.000000
OverallQual      0.790982
GrLivArea        0.708624
GarageCars       0.640409
GarageArea       0.623431
TotalBsmtSF      0.613581
1stFlrSF         0.605852
FullBath         0.560664
TotRmsAbvGrd     0.533723
YearBuilt        0.522897
YearRemodAdd     0.507101
GarageYrBlt      0.486362
MasVnrArea       0.477493
Fireplaces       0.466929
BsmtFinSF1       0.386420
LotFrontage      0.351799
WoodDeckSF       0.324413
2ndFlrSF         0.319334
OpenPorchSF      0.315856
HalfBath         0.284108
LotArea          0.263843
BsmtFullBath     0.227122
BsmtUnfSF        0.214479
BedroomAbvGr     0.168213
ScreenPorch      0.111447
PoolArea         0.092404
MoSold           0.046432
3SsnPorch        0.044584
BsmtFinSF2      -0.011378
BsmtHalfBath    -0.016844
MiscVal         -0.021190
Id              -0.021917
LowQualFinSF    -0.025606
YrSold          -0.028923
OverallCond     -0.077856
MSSubClass      -0.084284
EnclosedPorch   -0.128578
KitchenAbvGr    -0.135907
Name: SalePr

In [3]:
# There are 18 features that have a correlation coefficient greater than 0.3 with SalePrice
# Lets use the top 5 of those features

corr_features = ['OverallQual', 'GrLivArea', 'GarageCars', 'GarageArea', 'TotalBsmtSF']

# Lets also set the target variable
y = iowa_data.SalePrice
X = iowa_data[corr_features]
X.head()

OverallQual  GrLivArea  GarageCars  GarageArea  TotalBsmtSF
0            7       1710           2         548          856
1            6       1262           2         460         1262
2            7       1786           2         608          920
3            7       1717           3         642          756
4            8       2198           3         836         1145

In [4]:
# Lets split the data into training and testing sets
from sklearn.model_selection import train_test_split
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

In [5]:
# Lets get our model in and train it, we will be using random forest
from sklearn.ensemble import RandomForestRegressor

rf_model = RandomForestRegressor(random_state=1)
rf_model.fit(train_X, train_y)

RandomForestRegressor(random_state=1)

In [6]:

from sklearn.metrics import mean_absolute_error
# Now lets make predictions and evaluate the model
val_predictions = rf_model.predict(val_X)
val_mae = mean_absolute_error(val_y, val_predictions)
print(val_mae)

20618.935603228427


This a good start with some strongly correlated features.
Lets set up different number of features and also measure accuracy with root-mean-squared-error (RMSE)

In [7]:
# All of these features have at least a correlation coefficient greater than 0.3 with SalePrice

price_corr_matrix = corr_matrix["SalePrice"].sort_values(ascending=False)
corr_features5 = price_corr_matrix[1:6].index
corr_features10 = price_corr_matrix[1:11].index
corr_features15 = price_corr_matrix[1:16].index
corr_features18 = price_corr_matrix[1:19].index

In [8]:
from sklearn.metrics import root_mean_squared_error

X_5 = iowa_data[corr_features5]
X_10 = iowa_data[corr_features10]
X_15 = iowa_data[corr_features15]
X_18 = iowa_data[corr_features18]

# Method to take in features, the target and the model
def get_model_score(features, target, model):
    train_X, val_X, train_y, val_y = train_test_split(features, target, random_state=1)
    model.fit(train_X, train_y)
    predictions = model.predict(val_X)
    rmse = root_mean_squared_error(val_y, predictions)
    return rmse

rmse_5 = get_model_score(X_5, y, rf_model)
rms_10 = get_model_score(X_10, y, rf_model)
rms_15 = get_model_score(X_15, y, rf_model)
rms_18 = get_model_score(X_18, y, rf_model)
print(rmse_5, rms_10, rms_15, rms_18)

28270.259773953843 27961.11933058815 29206.551138146602 26668.85677760351


In [9]:
# Lets do a quick check with my feature picks: OverallQual, GrLivArea, GarageArea

X_3 = iowa_data[["OverallQual", "GrLivArea", "GarageArea"]]

rmse_3 = get_model_score(X_3, y, rf_model)
print(rmse_3)

# Also lets check with the features that have a positive correlation coefficient

corr_features29 = price_corr_matrix[1:28].index
X_29 = iowa_data[corr_features29]

rmse_29 = get_model_score(X_29, y, rf_model)
print(rmse_29)

32577.429467446957
27774.992909887456


We have selected 3, 5, 10, 15, 18, 29 different set of features, all which have a positive correlation to SalePrice
These are the results for different numbers of features.

| Feature Numbers | RMSE |
| --- | --- |
| 3 | 32577.43 |
| 5 | 28270.26 |
| 10 | 27961.12 |
| 15 | 29206.55 |
| 18 | 26668.86 |
| 29 | 27774.99 |

**The best seems to be 18 features**

## 2. Using RandomForest (Tree Based Models) for getting feature importance
Above in one of the cells we used and trained a random forest regressor model. We can use this model to get feature importance which is best for capturing non linear relationships and rank their importance

In [16]:
# Get feature importance
corr_features27 = price_corr_matrix[1:28].index
X_27 = iowa_data[corr_features27]
features_importance = pd.Series(rf_model.feature_importances_, index=X_27.columns)
print(features_importance.sort_values(ascending=False))

OverallQual     0.549712
GrLivArea       0.116866
TotalBsmtSF     0.043855
1stFlrSF        0.035069
BsmtFinSF1      0.033715
LotArea         0.027896
GarageCars      0.025777
GarageArea      0.020904
YearBuilt       0.017094
2ndFlrSF        0.015537
LotFrontage     0.014835
YearRemodAdd    0.013548
GarageYrBlt     0.012355
TotRmsAbvGrd    0.012118
MasVnrArea      0.010160
MoSold          0.009146
FullBath        0.008157
BsmtUnfSF       0.007850
WoodDeckSF      0.007190
OpenPorchSF     0.006940
Fireplaces      0.004061
ScreenPorch     0.002689
BsmtFullBath    0.001538
BedroomAbvGr    0.001416
HalfBath        0.000998
PoolArea        0.000323
3SsnPorch       0.000253
dtype: float64


**This shows the importance by percentage**
There are a few ways to pick out feature from these numbers.
1. **Eliminate** values that are low importance ( less than .01 ) or near zero
2. Use cumulative sum approach to get feature that get up 95% importance total
3. If the data has a lot of features, we can test around with picking an arbitrary cut off (ex. let use the top 10 most important feature)

Once we pick the features we can also sanity test by training with low importance features and then training without them and seeing how that affects accuracy

## 3. Detect multicollinearity
This is best to detect highly correlated features to reduce redundancy
We will be using Variance Inflation Factor

In [20]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

vif_data = pd.DataFrame()
vif_data["Feature"] = X_27.columns
vif_data["VIF"] = [variance_inflation_factor(X_27.values, i) for i in range(X_27.shape[1])]
print(vif_data.sort_values(by="VIF", ascending=False))

MissingDataError: exog contains inf or nans

In [21]:
import statsmodels.api as sm

# Add constant for intercept
X_const = sm.add_constant(X)

# Fit OLS Regression
model = sm.OLS(y, X_const).fit()

# Display Summary
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:              SalePrice   R-squared:                       0.761
Model:                            OLS   Adj. R-squared:                  0.760
Method:                 Least Squares   F-statistic:                     926.5
Date:                Mon, 17 Mar 2025   Prob (F-statistic):               0.00
Time:                        23:38:41   Log-Likelihood:                -17499.
No. Observations:                1460   AIC:                         3.501e+04
Df Residuals:                    1454   BIC:                         3.504e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -9.907e+04   4638.450    -21.359      